Bottom: 5834a8fd4aa7f8100a8091ac2c0b0fdf66c75aee
Top:    6960ac5323830409fc42b5512fc38415cf7983e3
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-14 15:38:56 -0500

Add codebase generator function

The codebase generator takes information about a change, like repository,
branch, and revision, and creates a predictable string describing the code.
This is used to decide upon the builder for a change and also to tie a fork
repository back to its parent for merge requests.

---
diff --git a/docker/buildbot-master/buildbot/master.cfg b/docker/buildbot-master/buildbot/master.cfg
index 4eb34ff2..5f1c821e 100644
--- a/docker/buildbot-master/buildbot/master.cfg
+++ b/docker/buildbot-master/buildbot/master.cfg
@@ -6,6 +6,7 @@
 
 import os
 from buildbot.plugins import secrets,steps,util
+from urlparse import urlparse
 
 os.environ['LONG_RUN_TIMEOUT']='14400'
 
@@ -139,3 +140,28 @@ c['workers'].extend(ptxdist.workers)
 c['schedulers'].extend(ptxdist.schedulers)
 c['builders'].extend(ptxdist.builders)
 c['change_source'].extend(ptxdist.change_source)
+
+
+def codebaseGenerator(chdict):
+    """ Generate a codebase description.
+    
+    The codebase generator takes information about a change, like repository,
+    branch, and revision, and creates a predictable string describing the code.
+    This is used to decide upon the builder for a change and also to tie a fork
+    repository back to its parent for merge requests.
+    """
+    pprint(chdict, indent=2)
+    path = urlparse(chdict['revlink']).path.lstrip('/')
+
+    components = []
+    while path != "":
+        path, tail = os.path.split(path)
+        if tail != "":
+            components.append(tail)
+    components.reverse()
+
+    codebase = os.path.join(*components[0:2])
+    return codebase
+
+
+c['codebaseGenerator'] = codebaseGenerator
